src/main/java/org/cactoos/collection/Immutable.java

```diff
@@ -54,8 +54,9 @@
      * Ctor.
      * @param src Source collection
      */
-    public Immutable(final Collection<X> src) {
-        this.col = src;
+    @SuppressWarnings("unchecked")
+    public Immutable(final Collection<? extends X> src) {
+        this.col = (Collection<X>) src;
```

Comment:
@rocket-3 instead of this, you should change the type of the `col` field. You won't need any cast in that case.

Reply:
@rocket-3 then you need to propagate the same kind of change everywhere needed until everything types correctly :) that's the purpose of this series of tickets: that the whole codebase uses generic with variance like this.

Reply:
@victornoel, sorry, I didn't get it 

As I see, we have two ways:

- to do unchecked cast
- to rely on class which does uchecked cast

We can't propagate changes to JDK

Reply:
@rocket-3 We should be able to propagate as much as possible in our classes, and then normally, no cast is needed. If some casts are needed, let's put them at the last moment possible (usually inside a method implementation).

So let's go with this and see what happens with the two classes you started modifying!

Reply:
@rocket-3 by the way, it looks like `iterator()` is incorrectly implemented here since it should return an immutable iterator. In that case, I think no casting will be needed

Reply:
@victornoel, no other way for `NoNulls`, cause we have an `add` and `addAll` impls.
We can't add items to a Collection<? extends X> by JDK design, only extract, see [PECS trilemma](https://stackoverflow.com/a/4343547/16625980). 

By the way, changed `Immutable`, see in the recent commit.

Reply:
@rocket-3 ok, in that case, for `NoNulls`, because of its nature (being mutable), it's not possible to apply variance like this and thus the constructor should only take a X. If we cast, this will create errors at runtime if incorrect elements are added to such a collection. Let's also add a comment in the javadoc of the constructor so that we remember why it's like this.

Reply:
@victornoel, wrote a test for this behaviour. Seems like no runtime exceptions, WDYT?

Reply:
@rocket-3 here is a failing test:
```java
    @Test
    void fails() {
        Collection<Double> a = new ArrayList<Double>();
        Collection<Number> b = new NoNulls<>(a);
        b.add(Long.valueOf(4));
        Double oupsNotADouble = a.iterator().next();
    }
```

This will throw `java.lang.ClassCastException: java.lang.Long cannot be cast to java.lang.Double` even though it compiles correctly.

Reply:
Okay, I've removed uchecked cast, added impl. note and removed test created recently.